hail/python/hail/methods/impex.py

```diff
@@ -1293,12 +1293,70 @@ def import_gen(path,
     -------
     :class:`.MatrixTable`
     """
-    path = wrap_to_list(path)
+    gen_table = import_lines(path, min_partitions)
+    sample_table = import_lines(sample_file)
     rg = reference_genome.name if reference_genome else None
+    contig_recoding = contig_recoding
     if contig_recoding is None:
-        contig_recoding = {}
-    return MatrixTable(ir.MatrixRead(ir.MatrixGENReader(
-        path, sample_file, chromosome, min_partitions, tolerance, rg, contig_recoding, skip_invalid_loci)))
+        contig_recoding = hl.empty_dict(hl.tstr, hl.tstr)
+    else:
+        contig_recoding = hl.dict(contig_recoding)
+
+    gen_table = gen_table.transmute(data=gen_table.text.split(' '))
+
+    if chromosome is None:
+        last_rowf_idx = 5
+        contig_holder = gen_table.data[0]
+    else:
+        last_rowf_idx = 4
+        contig_holder = chromosome
+
+    contig_holder = contig_recoding.get(contig_holder, contig_holder)
+
+    position = hl.int(gen_table.data[last_rowf_idx - 2])
+    alleles = hl.array([hl.str(gen_table.data[last_rowf_idx - 1]), hl.str(gen_table.data[last_rowf_idx])])
+    rsid = gen_table.data[last_rowf_idx - 3]
+    varid = gen_table.data[last_rowf_idx - 4]
+    if rg is None:
+        locus = hl.struct(contig=contig_holder, position=position)
+    else:
+        if skip_invalid_loci:
+            locus = hl.if_else(hl.is_valid_locus(contig_holder, position, rg),
+                               hl.locus(contig_holder, position, rg),
+                               hl.missing(hl.tlocus(rg)))
+        else:
+            locus = hl.locus(contig_holder, position, rg)
+
+    gen_table = gen_table.annotate(locus=locus, alleles=alleles, rsid=rsid, varid=varid)
+    gen_table = gen_table.annotate(entries=gen_table.data[last_rowf_idx + 1:].map(lambda x: hl.float64(x))
+                                   .grouped(3).map(lambda x: hl.struct(GP=x)))
+
+    sample_table_count = sample_table.count() - 2  # Skipping first 2 unneeded rows in sample file
+    gen_table = gen_table.annotate_globals(cols=hl.range(sample_table_count).map(lambda x: hl.struct(col_idx=x)))
+    if skip_invalid_loci:
+        gen_table = gen_table.filter(hl.is_defined(gen_table.locus))
+    mt = gen_table._unlocalize_entries('entries', 'cols', ['col_idx'])
+
+    sample_table = sample_table.tail(sample_table_count).add_index()
+    sample_table = sample_table.annotate(s=sample_table.text.split(' ')[0])
+    sample_table = sample_table.key_by(sample_table.idx)
+    mt = mt.annotate_cols(s=sample_table[hl.int64(mt.col_idx)].s)
+
+    mt = mt.annotate_entries(GP=hl.bind(lambda x: hl.if_else(hl.abs(1.0 - x) > tolerance,
```

Comment:
Separate comment: Don't use `x` in the lambda if it could have a better name. Let's say `gp_sum` or something instead.